### ISSUE TITLE
Promises?

### ISSUE THREAD

**Issue by GeoffreyPlitt** [errorception/redis-memoizer#3]
Is there a way to make this memoize promises instead of old-style async functions?

**Comment by STRML** [comment]
I use the following helpers:

```js
// Converts a cb function into a Promise as it memoizes
export function memoizeAsync<T>(func: CallbackContainingFunction<T>, ttl: number): () => Promise<T> {
  return Promise.promisify(memoize(func, ttl));
}

// Memoizes an existing promise
export function memoizePromise<T>(func: PromiseReturningFunction<T>, ttl: number): () => Promise<T> {
  // Memoizer only takes cb-accepting functions, so we have to convert then convert back
  const cbFunc = function() {
    return func.apply(this, arguments).nodeify(arguments[arguments.length - 1]);
  };
  return module.exports.memoizeAsync(cbFunc, ttl);
}
```

**Comment by willfarrell** [comment]
Maybe add this functionality into your fork @STRML?

**Comment by STRML** [comment]
I'd rather not depend on any particular promise implementation, and this is relatively easy to shim as above.

**Comment by willfarrell** [comment]
Fair enough, shim it is then.

**Comment by rakeshpai** [comment]
+1 to shimming or using something like Bluebird, sine it'll let you use both styles of APIs. If someone can PR a promises version, I'll be glad to release a major version upgrade.

**Comment by willfarrell** [comment]
I'm currently using (till I get around to refactoring):
```javascript
var Promise = require('bluebird');
var redisClient = require('redis').createClient({
	host:           'redis',
	return_buffers: true
});
var memoize = require('redis-memoizer')(redisClient, {});

module.exports = {
	callback: (func, ttl) => {
		return memoize(func, ttl);
	},

	// Converts a cb function into a Promise as it memoizes
	async:    (func, ttl) => {
		return Promise.promisify(memoize(func, ttl));
	},

	// Memoizes an existing promise
	promise: (func, ttl) => {
		// Memoizer only takes cb-accepting functions, so we have to convert then convert back
		const cbFunc = function () {	// arrow function will break arguments
			return func.apply(this, arguments).nodeify(arguments[arguments.length - 1]);
		};
		return Promise.promisify(memoize(cbFunc, ttl));
	}
};
```
wrapping the fork from @STRML

I would suggest:
1. Merge the fork from https://github.com/BitMEX/redis-memoizer <- lots of good work here #2.
2. Make @STRML a contributor <- #2.
3. the best implementation I could come up with is to make the first arg the wrapper type {callback,async,promise}:
```javascript
var redisClient = require('redis').createClient({
	host:           'redis',
	return_buffers: true
});
var memoize-promise = require('redis-memoizer')('promise', redisClient, {});
```
may require a major version bump.
4. Optionally you allow the user to choose the Promise engine, but bluebird is practically the industry standard at this point due to it's speed.

Thoughts?

**Comment by rakeshpai** [comment]
Sorry, I'm just getting to this now. I apologise that this project has been suffering from neglect. It's been a case of 'it works for me', and I didn't get PRs, so it just let it be the way it is. Still in use all over though. :)

* I'm now fully on-board the promises train, thanks to async/await support having landed in node. I'm even open to making a breaking version change, and dropping support for the callback-style altogether. (I'm still not a fan of the way code looks when using promises, but async/await, with its built-in error-handling looks just beautiful.)
* I only just got the chance to review @STRML's fork, and the changes are awesome. Looks like the fork is actively maintained too. I take the blame for not responding sooner. Sorry. 
* I'm thinking of going with native promises, rather than bluebird, mainly to keep dependencies down. From what I understand (and I could be wrong), the relevant perf improvement we'd get from bluebird is because they've avoided the `new Promise` constructor call, which, you know, sure it's a minor improvement, but it also deviates from the spec. Not sure how important that is, but I think I'd prefer the warm-fuzzies of sticking to the spec. I'm open to change this
* Found this other awesome spinoff project (https://github.com/pwmckenna/node-redis-locking-promise-memoizer) (hat tip, @pwmckenna), and I'm thinking of stealing ideas from there too. In particular, it uses a promisified version of redis-lock to improve on the cache-stampede handling. I recently promisified [redis-lock](https://github.com/errorception/redis-lock) itself (before I knew of [pwmckenna's promisified wrapper](https://github.com/pwmckenna/redis-lock-q)). Using it to prevent multiple calls sounds only obvious. Kicking myself for not thinking of it earlier.
